File: src/main/store/vault.ts
import { app, safeStorage } from 'electron'
import { join } from 'path'
import { randomBytes, scryptSync, createCipheriv, createDecipheriv, timingSafeEqual } from 'crypto'
import { readJsonVersioned, writeJsonVersioned } from './jsonStore'
import { FerroError } from '@shared/errors'
import { createLogger } from '../core/logger'

const log = createLogger('vault')

// Kimlik bilgisi şifreleme — iki mod:
//  • 'os'     : OS keychain'e bağlı safeStorage ( 'v1:' önekli ). Varsayılan.
//  • 'master' : Kullanıcı ana parolası; scrypt + AES-256-GCM ( 'm1:' önekli ).
//               OS keychain'e güvenmeyen/taşınabilir kurulumlar için.
// safeStorage YOKSA ve master ayarlı DEĞİLSE parola kalıcı SAKLANMAZ (null döner).
// Eski 'p0:' base64 kayıtları yalnızca OKUNUR (geri uyumluluk + demo seed).

const VAULT_META_VERSION = 1
const SCRYPT_N = 16_384
const SCRYPT_r = 8
const SCRYPT_p = 1
const KEY_LEN = 32
const IV_LEN = 12
const VERIFY_PLAINTEXT = 'ferro-vault-verify-v1'

interface MasterParams {
  /** scrypt tuzu (base64). */
  salt: string
  /** Doğrulayıcı: VERIFY_PLAINTEXT'in şifrelenmiş hali (base64 iv|tag|ct). */
  verifier: string
}

interface VaultMeta {
  mode: 'os' | 'master'
  master?: MasterParams
}

class Vault {
  private meta: VaultMeta | null = null
  /** Master mod açıkken bellekteki türetilmiş anahtar (kilit açılınca dolar). */
  private masterKey: Buffer | null = null

  private file(): string {
    return join(app.getPath('userData'), 'vault.json')
  }

  private load(): VaultMeta {
    if (this.meta) return this.meta
    this.meta = readJsonVersioned<VaultMeta>(this.file(), VAULT_META_VERSION, () => ({
      mode: 'os'
    }))
    return this.meta
  }

  private save(): void {
    writeJsonVersioned(this.file(), VAULT_META_VERSION, this.load())
  }

  private deriveKey(password: string, salt: Buffer): Buffer {
    return scryptSync(password, salt, KEY_LEN, { N: SCRYPT_N, r: SCRYPT_r, p: SCRYPT_p })
  }

  private aesEncrypt(key: Buffer, plain: string): string {
    const iv = randomBytes(IV_LEN)
    const cipher = createCipheriv('aes-256-gcm', key, iv)
    const ct = Buffer.concat([cipher.update(plain, 'utf8'), cipher.final()])
    const tag = cipher.getAuthTag()
    return Buffer.concat([iv, tag, ct]).toString('base64')
  }

  private aesDecrypt(key: Buffer, blob: string): string {
    const buf = Buffer.from(blob, 'base64')
    const iv = buf.subarray(0, IV_LEN)
    const tag = buf.subarray(IV_LEN, IV_LEN + 16)
    const ct = buf.subarray(IV_LEN + 16)
    const decipher = createDecipheriv('aes-256-gcm', key, iv)
    decipher.setAuthTag(tag)
    return Buffer.concat([decipher.update(ct), decipher.final()]).toString('utf8')
  }

  // ── Durum ──

  osEncryptionAvailable(): boolean {
    try {
      return safeStorage.isEncryptionAvailable()
    } catch {
      return false
    }
  }

  mode(): 'os' | 'master' {
    return this.load().mode
  }

  hasMaster(): boolean {
    return !!this.load().master
  }

  /** Master modda ve henüz kilit açılmamışsa true. OS modunda her zaman false. */
  isLocked(): boolean {
    return this.load().mode === 'master' && this.masterKey === null
  }

  // ── Şifrele / çöz ──

  /**
   * Düz metni saklanabilir string'e çevirir; saklanamıyorsa null döner
   * (OS keychain yok ve master ayarlı değil, ya da master kilitli).
   */
  encryptSecret(plain: string): string | null {
    const meta = this.load()
    if (meta.mode === 'master') {
      if (!this.masterKey) {
        log.warn('master parola kilitli — parola saklanamadı')
        return null
      }
      return 'm1:' + this.aesEncrypt(this.masterKey, plain)
    }
    if (this.osEncryptionAvailable()) {
      return 'v1:' + safeStorage.encryptString(plain).toString('base64')
    }
    log.warn('safeStorage kullanılamıyor — parola kalıcı olarak saklanmayacak')
    return null
  }

  /** Saklanan değeri çözer; çözülemezse boş string. */
  decryptSecret(stored: string): string {
    try {
      if (stored.startsWith('m1:')) {
        if (!this.masterKey) return '' // kilitli — çağıran parolayı sorar
        return this.aesDecrypt(this.masterKey, stored.slice(3))
      }
      if (stored.startsWith('v1:')) {
        return safeStorage.decryptString(Buffer.from(stored.slice(3), 'base64'))
      }
      if (stored.startsWith('p0:')) {
        return Buffer.from(stored.slice(3), 'base64').toString('utf8')
      }
    } catch (err) {
      log.error('parola çözülemedi', String(err))
    }
    return ''
  }

  // ── Master parola yönetimi ──

  /** Master parolayı doğrular (verifier üzerinden). */
  verifyMaster(password: string): boolean {
    const m = this.load().master
    if (!m) return false
    try {
      const key = this.deriveKey(password, Buffer.from(m.salt, 'base64'))
      const decrypted = this.aesDecrypt(key, m.verifier)
      const a = Buffer.from(decrypted)
      const b = Buffer.from(VERIFY_PLAINTEXT)
      return a.length === b.length && timingSafeEqual(a, b)
    } catch {
      return false // GCM auth hatası = yanlış parola
    }
  }

  /** Master parolayla depoyu açar (oturum boyunca bellekte tutulur). */
  unlock(password: string): boolean {
    const m = this.load().master
    if (!m) return false
    if (!this.verifyMaster(password)) return false
    this.masterKey = this.deriveKey(password, Buffer.from(m.salt, 'base64'))
    log.info('kimlik deposu açıldı (master)')
    return true
  }

  /**
   * Master parolayı ayarlar/değiştirir ve yeni anahtar üretir.
   * Çağıran (IPC handler), mevcut sırları önce çözüp sonra yeni anahtarla
   * yeniden şifrelemekten sorumludur (transitionSecrets deseni).
   * Değişiklikte mevcut parola doğrulanır.
   */
  setMaster(current: string | undefined, next: string): void {
    const meta = this.load()
    if (meta.master) {
      if (!current || !this.verifyMaster(current)) {
        throw new FerroError('AUTH_FAILED', 'Mevcut master parola hatalı')
      }
    }
    const salt = randomBytes(16)
    const key = this.deriveKey(next, salt)
    const verifier = this.aesEncrypt(key, VERIFY_PLAINTEXT)
    meta.mode = 'master'
    meta.master = { salt: salt.toString('base64'), verifier }
    this.masterKey = key
    this.save()
    log.info('master parola ayarlandı')
  }

  /** OS keychain moduna döner (master parolayı kaldırır). Mevcut parola gerekir. */
  disableMaster(current: string): void {
    const meta = this.load()
    if (!meta.master) return
    if (!this.verifyMaster(current)) {
      throw new FerroError('AUTH_FAILED', 'Master parola hatalı')
    }
    meta.mode = 'os'
    meta.master = undefined
    this.masterKey = null
    this.save()
    log.info('OS keychain moduna dönüldü')
  }
}

const vault = new Vault()

// Dışa açık ince API (mevcut çağrı yerleriyle uyumlu).
export function encryptionAvailable(): boolean {
  // Renderer'a "parola saklanabilir mi" bilgisi: master ayarlıysa (kilitliyken bile
  // açılınca saklanır) ya da OS keychain varsa true.
  return vault.mode() === 'master' || vault.osEncryptionAvailable()
}
export function encryptSecret(plain: string): string | null {
  return vault.encryptSecret(plain)
}
export function decryptSecret(stored: string): string {
  return vault.decryptSecret(stored)
}
export { vault }

File: src-tauri/src/vault.rs
//! Kimlik bilgisi şifreleme — `src/main/store/vault.ts` karşılığı, iki mod:
//!  • `os`     : OS keychain (keyring). Rust'ta rastgele bir veri anahtarı keychain'de
//!               tutulur; sırlar AES-256-GCM ile şifrelenir — önek `k1:`.
//!  • `master` : Kullanıcı ana parolası; scrypt(N=16384,r=8,p=1) + AES-256-GCM — önek
//!               `m1:`. Bu format mevcut şifreli verilerle BYTE-UYUMLUDUR (iv‖tag‖ct, base64).
//!
//! Yalnızca okunur geri uyumluluk:
//!  • `p0:` base64 düz metin (demo seed) — çözülür.
//!  • `v1:` eski OS anahtar zinciri (safeStorage) — Rust ÇÖZEMEZ; boş string döner (parola yeniden girilir).

use std::path::PathBuf;
use std::sync::Mutex;

use aes_gcm::aead::{Aead, KeyInit};
use aes_gcm::{Aes256Gcm, Key, Nonce};
use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine;
use rand::RngCore;
use serde::{Deserialize, Serialize};

use crate::error::{FerroError, FerroErrorCode, FerroResult};
use crate::store::json_store::{read_versioned, write_versioned, ReadOutcome};

const VAULT_META_VERSION: u32 = 1;
const SCRYPT_LOG_N: u8 = 14; // N = 16384
const SCRYPT_R: u32 = 8;
const SCRYPT_P: u32 = 1;
const KEY_LEN: usize = 32;
const IV_LEN: usize = 12;
const TAG_LEN: usize = 16;
const VERIFY_PLAINTEXT: &[u8] = b"ferro-vault-verify-v1";
const KEYRING_SERVICE: &str = "com.ferro.app";
const KEYRING_ACCOUNT: &str = "vault-data-key";

#[derive(Debug, Clone, Serialize, Deserialize)]
struct MasterParams {
    /// scrypt tuzu (base64).
    salt: String,
    /// Doğrulayıcı: VERIFY_PLAINTEXT'in şifreli hâli (base64 iv‖tag‖ct).
    verifier: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct VaultMeta {
    mode: String, // "os" | "master"
    #[serde(default, skip_serializing_if = "Option::is_none")]
    master: Option<MasterParams>,
}

impl Default for VaultMeta {
    fn default() -> Self {
        Self { mode: "os".into(), master: None }
    }
}

struct VaultState {
    meta: VaultMeta,
    /// Master modda kilit açıkken bellekteki türetilmiş anahtar.
    master_key: Option<[u8; KEY_LEN]>,
    /// OS modu veri anahtarı önbelleği (keyring'den bir kez okunur).
    os_key: Option<[u8; KEY_LEN]>,
}

pub struct Vault {
    file: PathBuf,
    state: Mutex<VaultState>,
}

// ── scrypt + AES-256-GCM yardımcıları ──────────────────────────────────────

fn scrypt_key(password: &str, salt: &[u8]) -> FerroResult<[u8; KEY_LEN]> {
    let params = scrypt::Params::new(SCRYPT_LOG_N, SCRYPT_R, SCRYPT_P, KEY_LEN)
        .map_err(|e| FerroError::with_detail(FerroErrorCode::Unknown, "scrypt parametresi", e.to_string()))?;
    let mut out = [0u8; KEY_LEN];
    scrypt::scrypt(password.as_bytes(), salt, &params, &mut out)
        .map_err(|e| FerroError::with_detail(FerroErrorCode::Unknown, "anahtar türetilemedi", e.to_string()))?;
    Ok(out)
}

/// Vault düzeni: base64(iv[12] ‖ tag[16] ‖ ct). aes-gcm çıktısı ct‖tag
/// olduğundan tag başa alınır.
fn aes_encrypt(key: &[u8; KEY_LEN], plain: &[u8]) -> FerroResult<String> {
    let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(key));
    let mut iv = [0u8; IV_LEN];
    rand::thread_rng().fill_bytes(&mut iv);
    let ct_tag = cipher
        .encrypt(Nonce::from_slice(&iv), plain)
        .map_err(|_| FerroError::new(FerroErrorCode::Unknown, "şifreleme başarısız"))?;
    let (ct, tag) = ct_tag.split_at(ct_tag.len() - TAG_LEN);
    let mut out = Vec::with_capacity(IV_LEN + TAG_LEN + ct.len());
    out.extend_from_slice(&iv);
    out.extend_from_slice(tag);
    out.extend_from_slice(ct);
    Ok(B64.encode(out))
}

fn aes_decrypt(key: &[u8; KEY_LEN], blob: &str) -> FerroResult<Vec<u8>> {
    let buf = B64
        .decode(blob)
        .map_err(|_| FerroError::new(FerroErrorCode::AuthFailed, "geçersiz base64"))?;
    if buf.len() < IV_LEN + TAG_LEN {
        return Err(FerroError::new(FerroErrorCode::AuthFailed, "şifreli veri eksik"));
    }
    let iv = &buf[..IV_LEN];
    let tag = &buf[IV_LEN..IV_LEN + TAG_LEN];
    let ct = &buf[IV_LEN + TAG_LEN..];
    // aes-gcm ct‖tag bekler.
    let mut ct_tag = Vec::with_capacity(ct.len() + TAG_LEN);
    ct_tag.extend_from_slice(ct);
    ct_tag.extend_from_slice(tag);
    let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(key));
    cipher
        .decrypt(Nonce::from_slice(iv), ct_tag.as_ref())
        .map_err(|_| FerroError::new(FerroErrorCode::AuthFailed, "şifre çözülemedi (yanlış parola?)"))
}

impl Vault {
    pub fn new(config_dir: &std::path::Path) -> Self {
        let file = config_dir.join("vault.json");
        let meta = match read_versioned::<VaultMeta, _>(&file, VAULT_META_VERSION, |_| None) {
            ReadOutcome::Loaded(m) => m,
            _ => VaultMeta::default(),
        };
        Self {
            file,
            state: Mutex::new(VaultState { meta, master_key: None, os_key: None }),
        }
    }

    fn save(&self, meta: &VaultMeta) {
        if let Err(e) = write_versioned(&self.file, VAULT_META_VERSION, meta) {
            eprintln!("[ferro] vault.json yazılamadı: {e}");
        }
    }

    /// OS modu veri anahtarını keychain'den okur; yoksa üretip saklar.
    fn os_data_key(&self, st: &mut VaultState) -> Option<[u8; KEY_LEN]> {
        if let Some(k) = st.os_key {
            return Some(k);
        }
        let entry = keyring::Entry::new(KEYRING_SERVICE, KEYRING_ACCOUNT).ok()?;
        let key = match entry.get_password() {
            Ok(b64) => {
                let bytes = B64.decode(b64).ok()?;
                if bytes.len() != KEY_LEN {
                    return None;
                }
                let mut k = [0u8; KEY_LEN];
                k.copy_from_slice(&bytes);
                k
            }
            Err(keyring::Error::NoEntry) => {
                let mut k = [0u8; KEY_LEN];
                rand::thread_rng().fill_bytes(&mut k);
                entry.set_password(&B64.encode(k)).ok()?;
                k
            }
            Err(_) => return None,
        };
        st.os_key = Some(key);
        Some(key)
    }

    // ── Durum ──

    pub fn os_encryption_available(&self) -> bool {
        // keyring erişilebiliyor mu (veri anahtarı okunabiliyor/üretilebiliyor mu)?
        let mut st = self.state.lock().unwrap();
        self.os_data_key(&mut st).is_some()
    }

    pub fn mode(&self) -> String {
        self.state.lock().unwrap().meta.mode.clone()
    }

    pub fn has_master(&self) -> bool {
        self.state.lock().unwrap().meta.master.is_some()
    }

    pub fn is_locked(&self) -> bool {
        let st = self.state.lock().unwrap();
        st.meta.mode == "master" && st.master_key.is_none()
    }

    /// Renderer'a "parola saklanabilir mi": master ayarlıysa ya da OS keychain varsa.
    pub fn encryption_available(&self) -> bool {
        self.mode() == "master" || self.os_encryption_available()
    }

    // ── Şifrele / çöz ──

    pub fn encrypt_secret(&self, plain: &str) -> Option<String> {
        let mut st = self.state.lock().unwrap();
        if st.meta.mode == "master" {
            let key = st.master_key?;
            return aes_encrypt(&key, plain.as_bytes()).ok().map(|b| format!("m1:{b}"));
        }
        let key = self.os_data_key(&mut st)?;
        aes_encrypt(&key, plain.as_bytes()).ok().map(|b| format!("k1:{b}"))
    }

    pub fn decrypt_secret(&self, stored: &str) -> String {
        let mut st = self.state.lock().unwrap();
        let result = if let Some(rest) = stored.strip_prefix("m1:") {
            match st.master_key {
                Some(key) => aes_decrypt(&key, rest).ok(),
                None => None, // kilitli — çağıran parolayı sorar
            }
        } else if let Some(rest) = stored.strip_prefix("k1:") {
            self.os_data_key(&mut st).and_then(|key| aes_decrypt(&key, rest).ok())
        } else if let Some(rest) = stored.strip_prefix("p0:") {
            B64.decode(rest).ok()
        } else {
            // v1: (eski OS anahtar zinciri) çözülemez.
            None
        };
        result.and_then(|b| String::from_utf8(b).ok()).unwrap_or_default()
    }

    // ── Master parola yönetimi ──

    pub fn verify_master(&self, password: &str) -> bool {
        let st = self.state.lock().unwrap();
        let Some(m) = &st.meta.master else { return false };
        let Ok(salt) = B64.decode(&m.salt) else { return false };
        let Ok(key) = scrypt_key(password, &salt) else { return false };
        match aes_decrypt(&key, &m.verifier) {
            Ok(pt) => pt == VERIFY_PLAINTEXT,
            Err(_) => false,
        }
    }

    pub fn unlock(&self, password: &str) -> bool {
        if !self.verify_master(password) {
            return false;
        }
        let mut st = self.state.lock().unwrap();
        let Some(m) = &st.meta.master else { return false };
        let Ok(salt) = B64.decode(&m.salt) else { return false };
        let Ok(key) = scrypt_key(password, &salt) else { return false };
        st.master_key = Some(key);
        true
    }

    /// Master parolayı ayarlar/değiştirir. Çağıran (handler) mevcut sırları önce
    /// çözüp sonra yeniden şifrelemekten sorumludur (transitionSecrets deseni).
    pub fn set_master(&self, current: Option<&str>, next: &str) -> FerroResult<()> {
        {
            let st = self.state.lock().unwrap();
            if st.meta.master.is_some() {
                drop(st);
                let ok = current.map(|c| self.verify_master(c)).unwrap_or(false);
                if !ok {
                    return Err(FerroError::new(FerroErrorCode::AuthFailed, "Mevcut master parola hatalı"));
                }
            }
        }
        let mut salt = [0u8; 16];
        rand::thread_rng().fill_bytes(&mut salt);
        let key = scrypt_key(next, &salt)?;
        let verifier = aes_encrypt(&key, VERIFY_PLAINTEXT)?;
        let mut st = self.state.lock().unwrap();
        st.meta.mode = "master".into();
        st.meta.master = Some(MasterParams { salt: B64.encode(salt), verifier });
        st.master_key = Some(key);
        let meta = st.meta.clone();
        drop(st);
        self.save(&meta);
        Ok(())
    }

    /// OS keychain moduna döner (master parolayı kaldırır). Mevcut parola gerekir.
    pub fn disable_master(&self, current: &str) -> FerroResult<()> {
        {
            let st = self.state.lock().unwrap();
            if st.meta.master.is_none() {
                return Ok(());
            }
        }
        if !self.verify_master(current) {
            return Err(FerroError::new(FerroErrorCode::AuthFailed, "Master parola hatalı"));
        }
        let mut st = self.state.lock().unwrap();
        st.meta.mode = "os".into();
        st.meta.master = None;
        st.master_key = None;
        let meta = st.meta.clone();
        drop(st);
        self.save(&meta);
        Ok(())
    }
}
